src/tbp/monty/frameworks/models/mixins/resampling_hypotheses_evidence.py

```diff
@@ -0,0 +1,454 @@
+        feature evidence. Specifically, it identifies the top-k nodes with the highest
+        evidence scores and samples hypotheses only from those nodes, making the process
+        more efficient than uniformly sampling from all graph nodes.
+
+        The sampling includes:
+          - Selecting the top-k node indices based on evidence scores, where k is
+             determined by the `informed_count` and the number of hypotheses per node.
+          - Fetching the 3D locations of only the selected top-k nodes.
+          - Generating rotations for each hypothesis using one of two strategies:
+            (a) If `initial_possible_poses` is set, rotations are uniformly sampled or
+                user-defined.
+            (b) Otherwise, alignments are computed between stored node poses and
+                sensed directions.
+
+        This targeted sampling improves efficiency by avoiding unnecessary computation
+        for nodes with low evidence, especially beneficial when informed sampling occurs
+        at every step.
+
+        Args:
+            features (dict): Input features.
+            graph_id (str): Identifier of the graph being queried.
+            informed_count (int): Number of fully informed hypotheses to sample.
+            input_channel: The channel for which hypotheses are sampled.
+
+        Returns:
+            Tuple[np.ndarray, np.ndarray, np.ndarray]: Tuple containing
+                selected locations, rotations, and evidence data.
+
+        """
+        # Return empty arrays for no hypotheses to sample
+        if informed_count == 0:
+            return np.zeros((0, 3)), np.zeros((0, 3, 3)), np.zeros(0)
+
+        num_hyps_per_node = self._num_hyps_per_node(features, input_channel)
+
+        # === Calculate selected evidence by top-k indices === #
+        if self.use_features_for_matching[input_channel]:
+            node_feature_evidence = self._calculate_feature_evidence_for_all_nodes(
+                features, input_channel, graph_id
+            )
+            # Find the indices for the nodes with highest evidence scores. The sorting
+            # is done in ascending order, so extract the indices from the end of
+            # the argsort array. We get the needed number of informed nodes not
+            # the number of needed hypotheses.
+            top_indices = np.argsort(node_feature_evidence)[
+                -int(informed_count // num_hyps_per_node) :
+            ]
+            node_feature_evidence_filtered = (
+                node_feature_evidence[top_indices] * self.feature_evidence_increment
+            )
+        else:
+            num_nodes = self.graph_memory.get_num_nodes_in_graph(graph_id)
+            top_indices = np.arange(num_nodes)[
+                : int(informed_count // num_hyps_per_node)
+            ]
+            node_feature_evidence_filtered = np.zeros(len(top_indices))
+
+        selected_evidence = np.tile(node_feature_evidence_filtered, num_hyps_per_node)
```

Comment:
Same with the code block here. I would make it more clear in the variable namings that this is different from the hypothesis evidence counts we are maintaining. Maybe just call it node_feature_match or best_match_nodes.